### ISSUE TITLE
ipa: netgroup htable uses different case for keys when storing and retrieving netgroups

### ISSUE THREAD

**Issue by pbrezina** [pbrezina/sssd-test#3111]
Cloned from Pagure issue: https://pagure.io/SSSD/sssd/issue/3116
- Created at 2016-08-02 10:50:12 by [mzidek](https://pagure.io/user/mzidek)
- Closed as Duplicate
- Assigned to [mzidek](https://pagure.io/user/mzidek)
---

When processing netgroups in the IPA provider, we use hash table to store all members. We lowercase the key (DN) when storing data to this hash table, but do not do the same when retrieving them.

This caused that member netgroups where never returned.

Steps to reproduce (with ipa provider):[[BR]]

1. have nested netgroups. For example (NETGROUP_NAME: MEMBER1, MEMBER2, ...):
    - ng1: user1
    - ng2: user2, ng1
2. getent netgroup ng2
3. check if ng2 is in chache (ldbedit)

Current results: ng 2 is not cached

### Comments

---

**Comment from [mzidek](https://pagure.io/user/mzidek) at [2016-08-02 10:50:40](https://pagure.io/SSSD/sssd/issue/3116#235534)**

Fields changed

owner: somebody => mzidek
patch: 0 => 1

---

**Comment from [mzidek](https://pagure.io/user/mzidek) at [2016-08-02 21:21:41](https://pagure.io/SSSD/sssd/issue/3116#235535)**

Could you provide a reproducer?
It would be very useful for writing a test.

cc:  => lslebodn

---

**Comment from [mzidek](https://pagure.io/user/mzidek) at [2016-08-02 21:22:35](https://pagure.io/SSSD/sssd/issue/3116#235536)**

I would say this ticket is a duplicate of #3117

---

**Comment from [mzidek](https://pagure.io/user/mzidek) at [2016-08-03 09:37:16](https://pagure.io/SSSD/sssd/issue/3116#235537)**

attachment
[0001-ipa_netgroups-Lowercase-key-to-htable.patch](/SSSD/sssd/issue/raw/files/23b4a04b44190aa20a80bf99ae92c7362ce0534987f6cc3eab7af9e00f338fd1-0001-ipa_netgroups-Lowercase-key-to-htable.patch)

---

**Comment from [mzidek](https://pagure.io/user/mzidek) at [2016-08-03 09:42:07](https://pagure.io/SSSD/sssd/issue/3116#235538)**

It is not duplicate of #3117. I added steps to reproduce and attached patch so that it is not lost in time/devel-list.

description: When processing netgroups in the IPA provider, we use hash table to store all members. We lowercase the key (DN) when storing data to this hash table, but do not do the same when retrieving them.

This caused that member netgroups where never returned. => When processing netgroups in the IPA provider, we use hash table to store all members. We lowercase the key (DN) when storing data to this hash table, but do not do the same when retrieving them.

This caused that member netgroups where never returned.

Steps to reproduce (with ipa provider):[[BR]]

1. have nested netgroups. For example (NETGROUP_NAME: MEMBER1, MEMBER2, ...):
 * ng1: user1
 * ng2: user2, ng1
2. getent netgroup ng2
3. check if ng2 is in chache (ldbedit)

Current results: ng 2 is not cached

---

**Comment from [mzidek](https://pagure.io/user/mzidek) at [2016-08-03 12:55:20](https://pagure.io/SSSD/sssd/issue/3116#235539)**

All data in sssd cache are internal things. Checking internal things should not be done with integration test.
If there is not a way how to test this without checking sssd cache then provided patch is not sufficient. There is missing an unit test.

If there is a way (even indirect) how to test this bug without checking internal data then please change steps to reproduce.

---

**Comment from [mzidek](https://pagure.io/user/mzidek) at [2016-08-03 13:46:53](https://pagure.io/SSSD/sssd/issue/3116#235540)**

I never said the steps to reproduce are for integration test. It just shows what the problem is (== netgroup is not stored in cache).

The bug described in #3117 is still present after applying this patch. As I already said on the devel list, I am Ok if the two tickets are pushed together, even though I do not see reason for it, because the bug described here is IMO obvious.

If you want to test it by other means then looking into the cache, then instead of looking inside the cache you can wait for negative cache to expire and run getent netgroup again. With this patch you will see correct results, without it you will see nothing.

Btw. I do not think it makes sense to write integration test for this ticket. Integration test for ticket #3117 would make more sense (if we had infrastructure for it).

---

**Comment from [mzidek](https://pagure.io/user/mzidek) at [2016-08-03 22:59:38](https://pagure.io/SSSD/sssd/issue/3116#235541)**

Thank you very much for confirmation that steps for reproducing are not for integration test.

We needn't have a separate ticket for each patch.
and the fix is required for proper resolving of nested netgroups. It's better to track it in single ticket; So in case of back-porting to older branches we will not miss it.

Closing as a duplicate of #3117

resolution:  => duplicate
status: new => closed

---

**Comment from [mzidek](https://pagure.io/user/mzidek) at [2017-02-24 14:52:09](https://pagure.io/SSSD/sssd/issue/3116#235542)**

**Metadata Update from @mzidek**:
- Issue assigned to mzidek
- Issue marked as blocked by: #3117
- Issue set to the milestone: NEEDS_TRIAGE
